### PR TITLE
feat: support computed attributes

### PR DESCRIPTION
**Pull request by aiwantaozi**
**Problem:**
Need computed attributes to merge the user input and schema default

**Solution:**
Implement computed attributes

**Related Issue:**
#1858

### CODE REVIEW

File: staging/utils/json/jsonpatch.go [hunk 1]
```diff
@@ -27,3 +27,27 @@ func PatchObject(obj, patchObject any) (any, error) {
 
 	return obj, nil
 }
+
+func ApplyPatches(patches ...[]byte) ([]byte, error) {
```
Comment (thxCode) [inline staging/utils/json/jsonpatch.go:31]: ApplyPatches(doc []byte, patches ...[]byte)

File: pkg/apis/resource/basic_view.go [hunk 2]
```diff
@@ -205,6 +202,14 @@ func (r *PatchRequest) Validate() error {
 		return err
 	}
 
+	// Set computedAttributes.
+	patched.EnvironmentID = r.Environment.Refer.ID()
```
Comment (thxCode) [inline pkg/apis/resource/basic_view.go:206]: 1. `r.Environment.Refer` cannot guarantee that it always receives an ID from the request, I think you can use `r.Environemnt.ID` or query with `patched` directly.
2. I saw we used `patched.EnvironmentID` previously, why do we need to change it here?

File: pkg/templates/openapi/traverse.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+func PostorderTraversal(root Node) []Node {
+	if root == nil {
+		return nil
+	}
+
+	var result []Node
+	stack := &Stack{}
+	stack.Push(root)
+
+	for len(stack.data) > 0 {
```
Comment (thxCode) [inline pkg/templates/openapi/traverse.go:13]: make `for len(stack.data) >0` to `for stack.Len() >0 `

File: pkg/templates/openapi/traverse.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+		node := stack.Pop()
+
+		result = append([]Node{node}, result...)
```
Comment (thxCode) [inline pkg/templates/openapi/traverse.go:16]: reduce slice extending.

File: pkg/apis/resource/helper.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		Where(environment.ID(entity.EnvironmentID)).
+		WithProject(func(pq *model.ProjectQuery) {
+			pq.Select(
+				project.FieldID,
+				project.FieldName)
+		}).
+		Only(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (thxCode) [inline pkg/apis/resource/helper.go:39]: can we reuse the input validated result?

Comment (aiwantaozi) [reply pkg/apis/resource/helper.go:39]: Need query project name here.

**Review by thxCode — APPROVED** [review]
LGTM

**Pull request merged**